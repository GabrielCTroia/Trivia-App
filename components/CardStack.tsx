import React, { FunctionComponent, useState } from "react";
import { ReactNode } from "react";
import { View, StyleSheet, Text, ViewProps } from "react-native";
import { Card } from "./Card";

const noop = (...args: any[]) => { };

export type CardStackProps = ViewProps & {
  // A collections of components and their ids to be rendered as cards
  cardWithIdMaps: {
    id: string;
    card: ReactNode
  }[];

  onSwipeLeft?: (cardId: string) => void,
  onSwipeRight?: (cardId: string) => void,
  onEnd?: () => void,
}

export const CardStack: FunctionComponent<CardStackProps> = ({ onSwipeLeft = noop, onSwipeRight = noop, onEnd = noop, ...props }) => {
  const [cardsLeftCount, setCardsLeftCount] = useState(props.cardWithIdMaps.length - 1);

  const onSwipe = (dir: 'left' | 'right', id: string) => {
    setCardsLeftCount((prev) => prev - 1);

    if (dir === 'left') {
      onSwipeLeft(id);
    } else {
      onSwipeRight(id);
    }

    if (cardsLeftCount === 0) {
      onEnd();
    }
  }

  return (
    <View style={[styles.container, props.style]}>
      {props.cardWithIdMaps.map(({ id, card }) => (
        <Card
          key={id}
          onSwipeLeft={() => onSwipe('left', id)}
          onSwipeRight={() => onSwipe('right', id)}
        >
          {card}
        </Card>)
      )}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {},
});
