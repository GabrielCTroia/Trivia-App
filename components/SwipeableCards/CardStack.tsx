import React, { useState } from "react";
import R from 'ramda';
import { View, StyleSheet, ViewProps } from "react-native";
import { Card } from "./Card";

const noop = (...args: any[]) => { };

export type CardStackProps<T extends { [key: string]: any } = {}> = ViewProps & {
  // A collections of components and their ids to be rendered as cards
  items: T[];
  keyExtractor: (item: T) => string;
  renderItem: (item: T) => React.ReactElement | null;

  onSwipeLeft?: (cardId: string) => void,
  onSwipeRight?: (cardId: string) => void,
  onEnd?: () => void,
}

export function CardStack<T>({ onSwipeLeft = noop, onSwipeRight = noop, onEnd = noop, ...props }: CardStackProps<T>) {
  const [cardsLeftCount, setCardsLeftCount] = useState(props.items.length - 1);

  const onSwipe = (dir: 'left' | 'right', id: string) => {
    setCardsLeftCount((prev) => prev - 1);

    if (dir === 'left') {
      onSwipeLeft(id);
    } else {
      onSwipeRight(id);
    }

    if (cardsLeftCount === 0) {
      onEnd();
    }
  }

  // Limit the number of items rendered at once
  const firstItems = R.take(2, props.items);

  return (
    <View style={[styles.container, props.style]}>
      {firstItems.reverse().map((item) => {
        const key = props.keyExtractor(item);
        return (
          <Card
            key={key}
            onSwipeLeft={() => onSwipe('left', key)}
            onSwipeRight={() => onSwipe('right', key)}
          >
            {props.renderItem(item)}
          </Card>
        )
      })}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {},
});
