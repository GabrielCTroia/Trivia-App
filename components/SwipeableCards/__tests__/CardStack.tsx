import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import { CardStack } from '../CardStack';
import { Text } from 'react-native';

it('renders correctly', () => {
  const items = [
    {
      id: 1,
      title: 'Item 1',
    },
    {
      id: 2,
      title: 'Item 1',
    },
    {
      id: 3,
      title: 'Item 1',
    }
  ]

  const tree = renderer.create(
    <CardStack
      items={items}
      keyExtractor={(item) => String(item.id)}
      renderItem={(item) => <Text>{item.title}</Text>}
    />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
