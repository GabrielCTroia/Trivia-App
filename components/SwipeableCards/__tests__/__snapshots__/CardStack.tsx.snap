// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {},
      undefined,
    ]
  }
>
  <View
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    style={
      Object {
        "height": "100%",
        "left": 0,
        "opacity": 1,
        "position": "absolute",
        "rotate": "0deg",
        "top": 0,
        "transform": Array [
          Object {
            "translateX": 0,
          },
          Object {
            "translateY": 0,
          },
        ],
        "width": "100%",
      }
    }
  >
    <Text>
      Item 1
    </Text>
  </View>
  <View
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    style={
      Object {
        "height": "100%",
        "left": 0,
        "opacity": 1,
        "position": "absolute",
        "rotate": "0deg",
        "top": 0,
        "transform": Array [
          Object {
            "translateX": 0,
          },
          Object {
            "translateY": 0,
          },
        ],
        "width": "100%",
      }
    }
  >
    <Text>
      Item 1
    </Text>
  </View>
</View>
`;
